deduce.go

```diff
@@ -690,6 +695,253 @@ func (sm *SourceMgr) deduceFromPath(path string) (deductionFuture, error) {
 	}, nil
 }
 
+type deductionCoordinator struct {
+	ctx      context.Context
```

Comment:
Okay! That makes a lot more sense. The root cause here is that you don't trust the caller to responsibly use the right `context` value.

So, the design tradeoff you're making is that you can be more confident that this code will work for callers who do weird things with the package (and using different contexts _would_ be weird), but you are sacrificing simplicity to gain that confidence. 

I don't think I agree that that's the right design decision here since gps isn't ever going to have, like, tons of direct users. It's likely that `dep` is the only user for quite a while. I think clear documentation on the source manager's constructor (use the same context, or children of the original context, whenever you call its methods) would be sufficient.

But! That's just my feeling. I think we've actually gotten to the core of things now, and we have the tradeoff much more clearly on the table, which is good enough for me. 👍